wallet/src/Pos/Wallet/Web/Tracking/Sync.hs

```diff
@@ -387,52 +366,60 @@ trackingRollbackTxs (getEncInfo -> encInfo) allAddress getDiff getTs txs =
-        let hh = headerHash blkHeader
+    rollbackTx CAccModifier{..} (tx, undo, blkHeader) = do
+        let wh@(WithHash _ txId) = withHash (taTx tx)
+            hh = headerHash blkHeader
             hhs = repeat hh
-            mDiff = getDiff blkHeader
-            mTs = getTs blkHeader
-            tx@(UnsafeTx (NE.toList -> inps) (NE.toList -> outs) _) = taTx
-            !txid = hash taTx
-            undoL' = catMaybes undoL
-            resolvedInputs = zip inps undoL'
-            txOutgoings = map txOutAddress outs
-            txInputs = map (toaOut . snd) resolvedInputs
-
-            ownInputs = selectOwnAddresses encInfo (txOutAddress . toaOut) undoL'
-            ownOutputs = selectOwnAddresses encInfo txOutAddress $ outs
-            ownInputMetas = map snd ownInputs
-            ownOutputMetas = map snd ownOutputs
-            ownInputAddrs = map cwamId ownInputMetas
-            ownOutputAddrs = map cwamId ownOutputMetas
-
-            l = fromIntegral (length outs) :: Word32
-            ownTxIns = zip inps $ map fst ownInputs
-            ownTxOuts = map (TxInUtxo txid) ([0 .. l - 1] :: [Word32])
-
-            th = THEntry txid tx mDiff txInputs txOutgoings mTs
+            (ownInputs, ownOutputs, thEntry) =
+                selectOwnTxInsAndOuts encInfo (wh, undo) (getDiff blkHeader, getTs blkHeader)
+
+            ownTxOutIns = map (fst . fst) ownOutputs
 
             deletedHistory =
-                if (not $ null ownInputAddrs) || (not $ null ownOutputAddrs)
-                then DL.snoc camDeletedHistory th
+                if (not $ null ownInputs) || (not $ null ownOutputs)
+                then DL.snoc camDeletedHistory thEntry
                 else camDeletedHistory
 
-            deletedPtxCandidates = DL.cons (txid, th) camDeletedPtxCandidates
+            deletedPtxCandidates = DL.cons (txId, thEntry) camDeletedPtxCandidates
 
         -- Rollback isn't needed, because we don't use @utxoGet@
         -- (undo contains all required information)
-        let usedAddrs = map cwamId ownOutputMetas
-            changeAddrs = evalChange allAddress ownInputAddrs ownOutputAddrs
+        let usedAddrs = map (cwamId . snd) ownOutputs
+            changeAddrs = evalChange allAddress (map (cwamId . snd) ownInputs) usedAddrs
         CAccModifier
-            (deleteAndInsertIMM ownOutputMetas [] camAddresses)
+            (deleteAndInsertIMM (map snd ownOutputs) [] camAddresses)
             (deleteAndInsertVM (zip usedAddrs hhs) [] camUsed)
             (deleteAndInsertVM (zip changeAddrs hhs) [] camChange)
-            (deleteAndInsertMM ownTxOuts ownTxIns camUtxo)
+            (deleteAndInsertMM ownTxOutIns (map fst ownInputs) camUtxo)
             camAddedHistory
             deletedHistory
             camAddedPtxCandidates
             deletedPtxCandidates
 
+type OwnTxInOuts = [((TxIn, TxOutAux), CWAddressMeta)]
+
+selectOwnTxInsAndOuts
```

Comment:
this function performs common logic of `trackingApplyTxs` and `trackingRollbackTxs` and being just refactoring